File: app/src/main/java/com/meiji/daily/module/zhuanlan/ZhuanlanView.java
package com.meiji.daily.module.zhuanlan;

import android.app.Application;
import android.arch.lifecycle.Observer;
import android.content.res.Resources;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.CardView;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.TypedValue;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.meiji.daily.App;
import com.meiji.daily.R;
import com.meiji.daily.bean.ZhuanlanBean;
import com.meiji.daily.binder.ZhuanlanViewBinder;
import com.meiji.daily.di.component.DaggerZhuanlanComponent;
import com.meiji.daily.di.module.ZhuanlanModule;
import com.meiji.daily.module.base.BaseFragment;
import com.meiji.daily.util.RecyclerViewUtil;
import com.meiji.daily.util.SettingUtil;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Named;

import me.drakeet.multitype.MultiTypeAdapter;


/**
 * Created by Meiji on 2017/11/29.
 */

public class ZhuanlanView extends BaseFragment implements SwipeRefreshLayout.OnRefreshListener {

    private static final String TAG = "ZhuanlanView";
    private static final String ARGUMENT_TYPE = "ARGUMENT_TYPE";
    @Inject
    ZhuanlanViewModel mModel;
    @Inject
    @Named("application")
    Application mApplication;
    private RecyclerView mRecyclerView;
    private SwipeRefreshLayout mSwipeRefreshLayout;
    private LinearLayout mRoot;
    private int mType;
    private MultiTypeAdapter mAdapter;

    public static ZhuanlanView newInstance(int type) {
        Bundle args = new Bundle();
        args.putInt(ARGUMENT_TYPE, type);
        ZhuanlanView fragment = new ZhuanlanView();
        fragment.setArguments(args);
        return fragment;
    }

    private void refreshUI() {
        Resources.Theme theme = getActivity().getTheme();
        TypedValue rootViewBackground = new TypedValue();
        TypedValue itemViewBackground = new TypedValue();
        TypedValue textColorPrimary = new TypedValue();
        theme.resolveAttribute(R.attr.rootViewBackground, rootViewBackground, true);
        theme.resolveAttribute(R.attr.itemViewBackground, itemViewBackground, true);
        theme.resolveAttribute(R.attr.textColorPrimary, textColorPrimary, true);
        mRoot.setBackgroundResource(rootViewBackground.resourceId);

        Resources resources = getResources();
        int childCount = mRecyclerView.getChildCount();
        for (int i = 0; i < childCount; i++) {
            CardView cardView = mRecyclerView.getChildAt(i).findViewById(R.id.cardview);
            cardView.setBackgroundResource(itemViewBackground.resourceId);

            TextView tv_name = cardView.findViewById(R.id.tv_name);
            tv_name.setTextColor(resources.getColor(textColorPrimary.resourceId));

            TextView tv_followersCount = cardView.findViewById(R.id.tv_followersCount);
            tv_followersCount.setTextColor(resources.getColor(textColorPrimary.resourceId));

            TextView tv_postsCount = cardView.findViewById(R.id.tv_postsCount);
            tv_postsCount.setTextColor(resources.getColor(textColorPrimary.resourceId));

            TextView tv_intro = cardView.findViewById(R.id.tv_intro);
            tv_intro.setTextColor(resources.getColor(textColorPrimary.resourceId));
        }

        RecyclerViewUtil.invalidateCacheItem(mRecyclerView);
    }

    @Override
    protected int attachLayoutId() {
        return R.layout.fragment_zhuanlan;
    }

    @Override
    protected void initViews(View view) {
        mRoot = view.findViewById(R.id.root);
        mRecyclerView = view.findViewById(R.id.recycler_view);
        mSwipeRefreshLayout = view.findViewById(R.id.refresh_layout);
        mRecyclerView.setHasFixedSize(true);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        // 设置下拉刷新的按钮的颜色
        mSwipeRefreshLayout.setColorSchemeColors(SettingUtil.getInstance().getColor());
        mSwipeRefreshLayout.setOnRefreshListener(this);
    }

    @Override
    protected void initData() {
        Bundle arguments = getArguments();
        if (arguments != null) {
            mType = arguments.getInt(ARGUMENT_TYPE);
        }
    }

    @Override
    protected void subscribeUI() {
        DaggerZhuanlanComponent.builder()
                .appComponent(App.sAppComponent)
                .zhuanlanModule(new ZhuanlanModule(this, mType, mApplication))
                .build().inject(this);

//        ZhuanlanViewModel.Factory factory = new ZhuanlanViewModel.Factory(App.sApp, mType);
//        mModel = ViewModelProviders.of(this, factory).get(ZhuanlanViewModel.class);
        mModel.getList().observe(this, new Observer<List<ZhuanlanBean>>() {
            @Override
            public void onChanged(@Nullable List<ZhuanlanBean> list) {
                if (null != list && list.size() > 0) {
                    onSetAdapter(list);
                } else {
                    onShowNetError();
                }
            }
        });
        mModel.isLoading().observe(this, new Observer<Boolean>() {
            @Override
            public void onChanged(@Nullable Boolean aBoolean) {
                if (aBoolean) {
                    onShowLoading();
                } else {
                    onHideLoading();
                }
            }
        });
        mModel.isRefreshUI().observe(this, new Observer<Boolean>() {
            @Override
            public void onChanged(@Nullable Boolean aBoolean) {
                refreshUI();
            }
        });
    }

    private void onSetAdapter(List<ZhuanlanBean> list) {
        if (mAdapter == null) {
            mAdapter = new MultiTypeAdapter(list);
            mAdapter.register(ZhuanlanBean.class, new ZhuanlanViewBinder());
            mRecyclerView.setAdapter(mAdapter);
        } else {
            mAdapter.notifyDataSetChanged();
        }
    }

    @Override
    public void onRefresh() {
        mModel.handleData();
    }

    private void onShowLoading() {
        mSwipeRefreshLayout.setRefreshing(true);
        mRecyclerView.setVisibility(View.GONE);
    }

    private void onHideLoading() {
        mSwipeRefreshLayout.setRefreshing(false);
        mRecyclerView.setVisibility(View.VISIBLE);
    }

    private void onShowNetError() {
        Snackbar.make(mSwipeRefreshLayout, R.string.network_error, Snackbar.LENGTH_SHORT).show();
        mSwipeRefreshLayout.setEnabled(true);
    }
}

File: app/src/main/java/com/meiji/daily/module/zhuanlan/ZhuanlanView.kt
package com.meiji.daily.module.zhuanlan

import android.arch.lifecycle.Observer
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.LinearLayoutManager
import android.util.TypedValue
import android.view.View
import com.meiji.daily.App
import com.meiji.daily.R
import com.meiji.daily.bean.ZhuanlanBean
import com.meiji.daily.binder.ZhuanlanViewBinder
import com.meiji.daily.module.base.BaseFragment
import com.meiji.daily.util.RecyclerViewUtil
import com.meiji.daily.util.SettingHelper
import kotlinx.android.synthetic.main.fragment_zhuanlan.*
import kotlinx.android.synthetic.main.item_zhuanlan.view.*
import me.drakeet.multitype.MultiTypeAdapter
import javax.inject.Inject


/**
 * Created by Meiji on 2017/11/29.
 */

class ZhuanlanView : BaseFragment(), SwipeRefreshLayout.OnRefreshListener {
    @Inject
    lateinit var mModel: ZhuanlanViewModel
    @Inject
    lateinit var mSettingHelper: SettingHelper

    private var mAdapter: MultiTypeAdapter? = null

    private fun refreshUI() {
        val theme = activity!!.theme
        val rootViewBackground = TypedValue()
        val itemViewBackground = TypedValue()
        val textColorPrimary = TypedValue()
        theme.resolveAttribute(R.attr.rootViewBackground, rootViewBackground, true)
        theme.resolveAttribute(R.attr.itemViewBackground, itemViewBackground, true)
        theme.resolveAttribute(R.attr.textColorPrimary, textColorPrimary, true)
        ll_root!!.setBackgroundResource(rootViewBackground.resourceId)

        val resources = resources
        val childCount = recycler_view!!.childCount
        for (i in 0 until childCount) {
            val cardView = recycler_view!!.getChildAt(i).cardview
            cardView.setBackgroundResource(itemViewBackground.resourceId)

            cardView.tv_name.setTextColor(resources.getColor(textColorPrimary.resourceId))
            cardView.tv_followersCount.setTextColor(resources.getColor(textColorPrimary.resourceId))
            cardView.tv_postsCount.setTextColor(resources.getColor(textColorPrimary.resourceId))
            cardView.tv_intro.setTextColor(resources.getColor(textColorPrimary.resourceId))
        }
        RecyclerViewUtil.invalidateCacheItem(recycler_view)
    }

    override fun initInject() {
        DaggerZhuanlanComponent.builder()
                .appComponent(App.sAppComponent)
                .zhuanlanModule(ZhuanlanModule(this))
                .build().inject(this)
    }

    override fun attachLayoutId() = R.layout.fragment_zhuanlan

    override fun initViews(view: View) {
        recycler_view.setHasFixedSize(true)
        recycler_view.layoutManager = LinearLayoutManager(activity)
        // 设置下拉刷新的按钮的颜色
        refresh_layout.setColorSchemeColors(mSettingHelper.color)
        refresh_layout.setOnRefreshListener(this)
    }

    override fun subscribeUI() {
        mModel.mList.observe(this, Observer<List<ZhuanlanBean>> { list ->
            if (null != list && list.size > 0) {
                onSetAdapter(list)
            } else {
                onShowNetError()
            }
        })
        mModel.isLoading.observe(this, Observer<Boolean> { aBoolean ->
            if (aBoolean!!) {
                onShowLoading()
            } else {
                onHideLoading()
            }
        })
        mModel.isRefreshUI.observe(this, Observer<Boolean> { refreshUI() })
    }

    private fun onSetAdapter(list: List<ZhuanlanBean>) {
        if (mAdapter == null) {
            mAdapter = MultiTypeAdapter(list)
            mAdapter?.register(ZhuanlanBean::class.java, ZhuanlanViewBinder())
            recycler_view.adapter = mAdapter
        } else {
            mAdapter?.notifyDataSetChanged()
        }
    }

    override fun onRefresh() {
        mModel.handleData()
    }

    private fun onShowLoading() {
        refresh_layout.isRefreshing = true
        recycler_view.visibility = View.GONE
    }

    private fun onHideLoading() {
        refresh_layout.isRefreshing = false
        recycler_view.visibility = View.VISIBLE
    }

    private fun onShowNetError() {
        Snackbar.make(refresh_layout, R.string.network_error, Snackbar.LENGTH_SHORT).show()
        refresh_layout.isEnabled = true
    }

    companion object {

        internal val ARGUMENT_TYPE = "ARGUMENT_TYPE"
        internal val TAG = "ZhuanlanView"

        fun newInstance(type: Int): ZhuanlanView {
            val args = Bundle()
            args.putInt(ARGUMENT_TYPE, type)
            val fragment = ZhuanlanView()
            fragment.arguments = args
            return fragment
        }
    }
}
